v-next/hardhat-viem/src/contracts.ts

```diff
@@ -0,0 +1,233 @@
+    getContractAbiAndBytecode(artifactManager, contractName, libraries),
+  ]);
+
+  let deploymentTxHash: PrefixedHexString;
+  // If gasPrice is defined, then maxFeePerGas and maxPriorityFeePerGas
+  // must be undefined because it's a legaxy tx.
+  if (deployContractParameters.gasPrice !== undefined) {
+    deploymentTxHash = await walletClient.deployContract({
+      abi,
+      bytecode,
+      args: constructorArgs,
+      ...deployContractParameters,
+      maxFeePerGas: undefined,
+      maxPriorityFeePerGas: undefined,
+    });
+  } else {
+    deploymentTxHash = await walletClient.deployContract({
+      abi,
+      bytecode,
+      args: constructorArgs,
+      ...deployContractParameters,
+      gasPrice: undefined,
+    });
+  }
+
+  const deploymentTx = await publicClient.getTransaction({
+    hash: deploymentTxHash,
+  });
+
+  const contractAddress = getContractAddress({
+    from: walletClient.account.address,
+    nonce: toBigInt(deploymentTx.nonce),
+  });
+
+  const contract = createContractInstance(
+    contractName,
+    publicClient,
+    walletClient,
+    abi,
+    contractAddress,
+  );
+
+  return { contract, deploymentTransaction: deploymentTx };
+}
+
+export async function getContractAt<ContractName extends string>(
+  provider: EthereumProvider,
+  artifactManager: ArtifactsManager,
+  contractName: ContractName,
+  address: viemT.Address,
+  getContractAtConfig: GetContractAtConfig = {},
+): Promise<ContractReturnType<ContractName>> {
+  const [publicClient, walletClient, artifact] = await Promise.all([
+    getContractAtConfig.client?.public ?? getPublicClient(provider, "l1"),
+    getContractAtConfig.client?.wallet ??
+      getDefaultWalletClient(provider, "l1"),
+    artifactManager.readArtifact(contractName),
+  ]);
+
+  return createContractInstance(
+    contractName,
+    publicClient,
+    walletClient,
+    artifact.abi,
+    address,
+  );
+}
+
+function createContractInstance<ContractName extends string>(
+  _contractName: ContractName,
+  publicClient: PublicClient,
+  walletClient: WalletClient,
+  abi: viemT.Abi,
+  address: viemT.Address,
+): ContractReturnType<ContractName> {
+  const contract = getContract({
+    address,
+    client: {
+      public: publicClient,
+      wallet: walletClient,
+    },
+    abi,
+  });
+
+  /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+  -- Cast it as TS can't infer the type of the contract */
+  return contract as ContractReturnType<ContractName>;
+}
+
+async function getContractAbiAndBytecode(
+  artifacts: ArtifactsManager,
+  contractName: string,
+  libraries: Libraries,
+) {
+  const artifact = await artifacts.readArtifact(contractName);
+  let bytecode;
+  try {
+    bytecode = resolveLinkedBytecode(artifact, libraries);
+  } catch (error) {
+    ensureError(error);
+
+    throw new HardhatError(HardhatError.ERRORS.VIEM.LINKING_CONTRACT_ERROR, {
+      contractName,
+      error: error.message,
```

Comment:
The actual error, along with the list of offending libraries, comes from `resolveLinkedBytecode`. I think centralizing the error format is a good approach, but another option could be returning an object containing the validation errors and formatting them within the consumer plugin.